### PR TITLE
Restore performance in JoC RN50 inference.

### PR DESCRIPTION
**Pull request by mzient**
* Change synchronization point from transfer to decoding.
* Reduce number of events.
* Remove unnecessary synchronization with main stream.
* Change construction/destruction to use ranged for loops.
* Change .cu to .cc for faster builds and better diagnostics.

Signed-off-by: Michal Zientkiewicz <michalz@nvidia.com>

### CODE REVIEW

File: dali/pipeline/pipeline.cc [hunk 1]
```diff
@@ -182,16 +182,16 @@ void Pipeline::AddOperator(OpSpec spec, const std::string& inst_name) {
   this->op_specs_for_serialization_.push_back(make_pair(inst_name, spec));
 
   // If necessary, split nvJPEGDecoder operator in two separated stages (CPU and Mixed-GPU)
-#ifdef NVJPEG_DECOUPLED_API
```
Comment (mzient) [inline dali/pipeline/pipeline.cc]: This change allows explicitly disabling `NVJPEG_DECOUPLED_API` in `dali/pipeline/operators/decoder/CMakeLists.txt`, without affecting the whole project - faster build when switching between APIs.

**Comment by mzient** [comment]
!build

File: dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h [hunk 2]
```diff
@@ -91,45 +93,64 @@ class nvJPEGDecoder : public Operator<MixedBackend>, CachedDecoderImpl {
...
-      CUDA_CALL(cudaStreamCreateWithPriority(&streams_[i], cudaStreamNonBlocking,
-                                             default_cuda_stream_priority_));
+    for (auto &buffer : pinned_buffers_) {
+      NVJPEG_CALL(nvjpegBufferPinnedCreate(handle_, nullptr, &buffer));
+    }
+    for (auto &buffer : device_buffers_) {
+      NVJPEG_CALL(nvjpegBufferDeviceCreate(handle_, nullptr, &buffer));
+    }
+    for (auto &stream : streams_) {
+      CUDA_CALL(cudaStreamCreateWithFlags(&stream, cudaStreamNonBlocking));
```
Comment (JanuszL) [inline dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h]: Don't remove priority.

File: dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h [hunk 3]
```diff
@@ -91,45 +93,64 @@ class nvJPEGDecoder : public Operator<MixedBackend>, CachedDecoderImpl {
...
     }
-    CUDA_CALL(cudaEventCreate(&master_event_));
   }
 
   ~nvJPEGDecoder() override {
     try {
       thread_pool_.WaitForWork();
-    } catch (...) {
-      // As other images being process might fail, but we don't care
+    } catch (const std::runtime_error &) {
```
Comment (JanuszL) [inline dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h]: Can we print the content of `std::runtime_error` ?

File: dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h (hunk 2 above)
Comment (mzient) [reply dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h]: (Un)done.

File: dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h (hunk 3 above)
Comment (mzient) [reply dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h]: Done.

File: dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h [hunk 4]
```diff
@@ -91,45 +93,67 @@ class nvJPEGDecoder : public Operator<MixedBackend>, CachedDecoderImpl {
...
+    }
   }
 
   ~nvJPEGDecoder() override {
     try {
       thread_pool_.WaitForWork();
-    } catch (...) {
-      // As other images being process might fail, but we don't care
+    } catch (const std::runtime_error &e) {
+      std::cerr << "An error occurred in nvJPEG worker thread:\n"
```
Comment (jantonguirao) [inline dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h]: use `DALI_WARN`

**Pull request merged**